src/main/java/org/jabref/gui/mergeentries/FetchAndMergeEntry.java

```diff
@@ -70,38 +70,143 @@ public void fetchAndMerge(BibEntry entry, Field field) {
+    }
+
+    private void processFetchedEntry(Optional<BibEntry> fetchedEntry, Field field, String fieldContent, BibEntry originalEntry) {
+        if (fetchedEntry.isPresent()) {
+            ImportCleanup cleanup = ImportCleanup.targeting(bibDatabaseContext.getMode(), preferences.getFieldPreferences());
+            cleanup.doPostCleanup(fetchedEntry.get());
+            mergeWithoutDialog(originalEntry, fetchedEntry.get());
+        } else {
+            dialogService.notify(Localization.lang("Cannot get info based on given %0: %1", field.getDisplayName(), fieldContent));
+        }
+    }
+
+    private void handleFetchException(Exception exception, IdBasedFetcher fetcher, BibEntry entry) {
+        if (hasAnySupportedField(entry)) {
+            LOGGER.error("Error while fetching bibliographic information", exception);
+            String fetcherName = fetcher.getName();
+            if (exception instanceof FetcherClientException) {
+                dialogService.showInformationDialogAndWait(Localization.lang("Fetching information using %0", fetcherName), Localization.lang("No data was found for the identifier"));
+            } else if (exception instanceof FetcherServerException) {
+                dialogService.showInformationDialogAndWait(Localization.lang("Fetching information using %0", fetcherName), Localization.lang("Server not available"));
             } else {
-                dialogService.notify(Localization.lang("No %0 found", field.getDisplayName()));
+                dialogService.showInformationDialogAndWait(Localization.lang("Fetching information using %0", fetcherName), Localization.lang("Error occurred %0", exception.getMessage()));
+            }
+        }
+    }
+
+    private void mergeWithoutDialog(BibEntry originalEntry, BibEntry fetchedEntry) {
+        NamedCompound ce = new NamedCompound(Localization.lang("Merge entry without user interaction"));
+
+        // Set of fields present in both the original and fetched entries
+        Set<Field> jointFields = new TreeSet<>(Comparator.comparing(Field::getName));
+        jointFields.addAll(fetchedEntry.getFields());
+        Set<Field> originalFields = new TreeSet<>(Comparator.comparing(Field::getName));
+        originalFields.addAll(originalEntry.getFields());
+
+        boolean edited = false;
+
+        // Compare entry type and update if needed
+        EntryType oldType = originalEntry.getType();
+        EntryType newType = fetchedEntry.getType();
+
+        if (!oldType.equals(newType)) {
+            originalEntry.setType(newType);
+            ce.addEdit(new UndoableChangeType(originalEntry, oldType, newType));
+            edited = true;
+        }
+
+        // Compare fields and set the longer value
+        for (Field field : jointFields) {
+            Optional<String> originalString = originalEntry.getField(field);
+            Optional<String> fetchedString = fetchedEntry.getField(field);
+
+            if (fetchedString.isPresent()) {
```

Comment:
Let's [reduce nesting and return early](https://craftbettersoftware.com/p/clean-code-7-tips-to-write-clean#:~:text=Reduce%20nesting%20with%20a%20return%20early) here.